### ISSUE TITLE
Unable to update app on Chromecast with Google TV

### ISSUE THREAD

**Issue by theschism** [yuliskov/SmartTubeNext#793]
I really love this app, and it works fine on my fire stick and I can install the APK find on a Chromecast Android TV v10 with Apr 5 2021 security patch.
     The issue is the built in upgrade feature will not let me update the application. I have set the application up in the OS to allow unknown sources but the update process seems to get caught up in a loop, no matter if I allow the app to install unknown sources or not. The OS will not let SmartTubeNext install a new APK.
     This may be an issue with Android TV v10 Apr 5 2021 update but I could not find a bug to validate it. 
     AFTVnews Downlonloader works to install apps from their app on the same device with the same process.
     I am a huge fan and only trying to alert you that this is happening. Thank you!

**Comment by dseven** [comment]
I have the same issue with self-update on my CCwGTV. I've been using adb to load the new APKs as they come out. It'd be more convenient if self-update could work. I'm not sure if the issue is with the OS or the app.

**Comment by theschism** [comment]
Cool workaround. I can't seem to find anything past 11 r22 (in the github release section). I was going to look through the code to find the update sites but haven't had time, I have just been using it with firetv (the only reason I use fire tv at this point) :-P

**Comment by dseven** [comment]
Latest builds are available via Telegram: https://t.me/SmartTubeNext_en

**Comment by panosru** [comment]
I'm facing the same issue, unfortunately, in the telegram channel, the latest beta builds aren't available.

**Comment by dseven** [comment]
I've found that NordVPN has the same issue on my CCwGTV (same error when it tries to update itself). It appears to be an operating system issue, not an issue with the app(s)

**Comment by panosru** [comment]
For those who run into that issue, check if you have (or had) a work account linked with your device.

You can identify if an application is linked to a work account by the small briefcase icon next to the application icon, for example:
![photo_2021-10-05 12 14 42](https://user-images.githubusercontent.com/400362/135995628-0e673f9c-d0db-41d3-b3ad-b3c5a34d0907.jpeg)

To make sure that you don't have any secondary accounts, run `adb shell` and from within the shell run `dumpsys user`, if the output looks like this:

![image](https://user-images.githubusercontent.com/400362/135996049-c05e5d0d-05bf-4bae-a4c9-e93df8dffc6c.png)

Then you have more than one account attached (it might be either directly by you or via 3rd party app). My advice is to remove all the accounts and start fresh because it seems that down the line you messed something up, at least this is what I did.

A normal output should look like this:

![image](https://user-images.githubusercontent.com/400362/135996400-9c240e1a-88da-4fc0-b5bc-7adb860c3fb8.png)

If you reset your device and your `dumpsys user` output looks like the second screenshot, then you should be good to go and the apps installed via `adb install` should be working fine.

If you don't want to reset your device, then you have two options.

### Option 1
Install the `Downloader` app, then instead of running `adb install`, run `adb push your.apk /sdcard/Downloader` and then from within the Downloader app, go to Files and select the app you want to install from the list, it will be attached to your account properly.

### Option 2
From the `dumpsys user` command, you can see the users of the device, for example, from the screenshot I posted you can see I have `UserInfo{0:null:13} serialNo=0` (user code: 13) and `UserInfo{10:new_user:30} serialNo=10` (user code: 30), one of them is the right one, but, when you manually install an app over a USB cable to a device which has a work profile, the app is installed on both the personal and the work profile, that is the default ADB behaviour.

If you don't want to install the `Downloader` app, then you will still have to push your apk file to your device and install it from within the device. For example, you can push your apk to /sdcard/ like so: `adb push your.apk /sdcard/` and then run `adb shell` to enter the shell of the device, from within the shell, if you want to install the app for the user 13, you run `pm install --user 13 /sdcard/your.apk`, then don't forget to run `rm -f /sdcard/your.apk` as it is no longer needed.

**Comment by panosru** [comment]
For those who run into that issue, check if you have (or had) a work account linked with your device.

You can identify if an application is linked to a work account by the small briefcase icon next to the application icon, for example:
![photo_2021-10-05 12 14 42](https://user-images.githubusercontent.com/400362/135995628-0e673f9c-d0db-41d3-b3ad-b3c5a34d0907.jpeg)

I noticed a strange behaviour in Google TV that does not seem to exist in other android devices that I own (smartphones and tablets). On a fresh device, when you enable developer mode and enable USB debugging if you enter the shell (`adb shell`) and run `pm list users` before running `adb install` to that device, then you will have only one user, like this:
![image](https://user-images.githubusercontent.com/400362/136006262-5a5be3a9-8261-4747-9a73-c38e96a887b4.png)

The problem now is that in Google TV, when you run `adb install` and install an app, a new user is created, and now you have two users, that behaviour isn't happening on my other android devices. `pm list users` after running `adb install`:
![image](https://user-images.githubusercontent.com/400362/135998963-3644370d-d727-410f-810e-d6708ccb47da.png)

The _Owner_ is your main account, the _new_user_ appears when you first install an app via adb, which isn't normal.

After a few system reset in Google TV, I'm not facing the above-described issue anymore, but I'm not sure how and why it appeared in the first place.

If you end up having that extra user and you don't want to reset your device, you can uninstall the app that has the work account icon on it and then install the `Downloader` app, then instead of running `adb install`, run `adb push your.apk /sdcard/Downloader` and then from within the Downloader app, go to Files and select the app you want to install from the list, it will be attached to your account properly.

If I'll have any new info on the matter, I will update my post.

**Comment by yolking** [comment]
I've had same problem and well I tried obvious solution based on @panosru reply - I just ran `pm remove-user ID`, where ID is first number after parenthesis for new user (10 in previous reply and it was same number in my case). This new_user gone, no more briefcases. Had to give permissions again, but update immediately worked like a charm even without restart. Thank you @panosru  for the info.

**Comment by nitram777** [comment]
Thanks! It was hard to find this thread, but finally I can confirm removing user 10 did solve the update issue on my CCwGTV.

I would guess, it´s OS related as I installed same version of SmartTubeNext on the same day with the same process on my CCwGTV and my Sony TV (Android TV 9 based). And the updating on Sony works without issues.

But thanks to your findings it´s now also updating on the Google dongle 👍

**Comment by Tommystus** [comment]
Yup, lucky I found this thread.  Issue just as described by panosru and removing the user ID 10 fixed it.

**Comment by cybrnook** [comment]
Saved my bacon on a Sony TV running Google TV os. ID was also 10.
